backup-stores/gcs/src/main/java/io/camunda/zeebe/backup/gcs/GcsBackupStore.java

```diff
@@ -7,34 +7,122 @@
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.BucketInfo;
 import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.Storage.BlobTargetOption;
+import com.google.cloud.storage.StorageException;
 import com.google.cloud.storage.StorageOptions;
 import io.camunda.zeebe.backup.api.Backup;
 import io.camunda.zeebe.backup.api.BackupIdentifier;
 import io.camunda.zeebe.backup.api.BackupIdentifierWildcard;
 import io.camunda.zeebe.backup.api.BackupStatus;
 import io.camunda.zeebe.backup.api.BackupStatusCode;
 import io.camunda.zeebe.backup.api.BackupStore;
+import io.camunda.zeebe.backup.common.BackupStatusImpl;
 import io.camunda.zeebe.backup.gcs.GcsBackupStoreException.ConfigurationException.BucketDoesNotExistException;
+import io.camunda.zeebe.backup.gcs.GcsBackupStoreException.UnexpectedManifestState;
+import io.camunda.zeebe.backup.gcs.manifest.Manifest;
+import java.io.IOException;
 import java.nio.file.Path;
 import java.util.Collection;
+import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
 
 public final class GcsBackupStore implements BackupStore {
+  public static final ObjectMapper MAPPER =
+      new ObjectMapper()
+          .registerModule(new Jdk8Module())
+          .registerModule(new JavaTimeModule())
+          .disable(WRITE_DATES_AS_TIMESTAMPS)
+          .setSerializationInclusion(Include.NON_ABSENT);
+
   private final GcsBackupConfig config;
+  private final Storage client;
+  private final BucketInfo bucketInfo;
+  private final Executor executor;
 
   public GcsBackupStore(final GcsBackupConfig config) {
+    this(config, buildClient(config));
+  }
+
+  public GcsBackupStore(final GcsBackupConfig config, final Storage client) {
     this.config = config;
+    this.client = client;
+    bucketInfo = BucketInfo.of(config.bucketName());
+    executor = Executors.newWorkStealingPool(4);
   }
 
   @Override
   public CompletableFuture<Void> save(final Backup backup) {
-    throw new UnsupportedOperationException();
+    return CompletableFuture.runAsync(() -> saveSync(backup), executor);
+  }
+
+  private void saveSync(final Backup backup) {
+    final var inProgress = Manifest.create(backup);
+    final var blobInfo = manifestBlobInfo(backup.id());
+    final Blob initial;
+
+    try {
+
+      initial =
+          client.create(
+              blobInfo, MAPPER.writeValueAsBytes(inProgress), BlobTargetOption.doesNotExist());
+    } catch (final StorageException e) {
+      throw new UnexpectedManifestState(
+          "Manifest for backup %s already exists".formatted(backup.id()));
```

Comment:
:wrench: lets try to follow our error message guidelines https://github.com/camunda/zeebe/wiki/Error-Guidelines